reconcile/gitlab_owners.py

```diff
@@ -68,20 +67,27 @@ def get_change_owners_map(self):
 
     def get_lgtms(self):
         """
-        Collects the usernames of all the '/lgtm' comments.
+        Collects the usernames of all valid '/lgtm' comments.
         """
         lgtms = []
         comments = self.gitlab.get_merge_request_comments(self.mr.iid)
-        for comment in comments:
+        current_diff = self.gitlab.get_merge_request_diffs(self.mr.iid)
 
+        for comment in comments:
             # Only interested in '/lgtm' comments
             if comment['body'] != '/lgtm':
                 continue
 
-            # Only interested in comments created after the top commit
-            # creation time
-            comment_created_at = dateparser.parse(comment['created_at'])
-            if comment_created_at < self.top_commit_created_at:
+            user = comment['user']
+            comment_id = comment['id']
+            state_key = generate_state_key(self.mr, comment)
+
+            try:
+                diff = self.state.get(key=state_key)
+                decoded_diff = json.loads(diff)
+                if current_diff != decoded_diff:
+                    continue
+            except:
```

Comment:
@maorfr How we're doing it right now is using the commit creation time-stamp of the top most commit. While this approach will work when there are legitimate code changes after the last lgtm, the problem arises when the bot needs to auto-rebase an MR (because when such a rebase happens, the top commits are re-created, causing the top commit's creation time stamp to change). Even though the rebase doesn't introduce any code changes, the lgtm is simply discarded because the commits were re-created.

The challenge here is to have the integration differentiate between a legitimate code change v/s an auto-rebase (currently the bot sees both the scenarios as the same thing) so that approvers don't have to issue redundant lgtms just because there was a bot rebase.

> if a label should be added or not just by checking if there are no new code changes after the last valid lgtm?

After discussing this idea with @apahim , we decided to try out using state. How this PR checks if new changes are real changes (or just bot rebase), is that each time there is a lgtm, the diff (which this lgtm is approving) is stored as a state. At the time of decided if a label should be added, the current diff is compared with the saved diff. Now the current diff will only change if there is a real code change (and not just a bot rebase).

Do you have any ideas on how we can achieve this (differentiating between a bot rebase and legitimate changes) without creating a mapping between LGTM -> diff?

We've discussed some ideas on [this](https://docs.google.com/document/d/1yJmgDM9PNQfwMHifn4mxENsj5hsRvEx6Do5yK4iFzoQ/edit?usp=sharing) doc that do not require using State. Let me know your thoughts. :)